staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/marshal.go

```diff
@@ -59,6 +60,27 @@ func (s *JSONSchemaPropsOrBool) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+func (s JSONSchemaPropsOrBool) MarshalCBOR() ([]byte, error) {
+	if s.Schema != nil {
+		return cbor.Marshal(s.Schema)
+	}
+	return cbor.Marshal(s.Allows)
+}
+
+func (s *JSONSchemaPropsOrBool) UnmarshalCBOR(data []byte) error {
+	var b bool
+	if err := cbor.Unmarshal(data, &b); err == nil {
+		*s = JSONSchemaPropsOrBool{Allows: b}
+		return nil
+	}
+	var p JSONSchemaProps
+	if err := cbor.Unmarshal(data, &p); err != nil {
+		return err
```

Comment:
At this point we have two errors, the one from a failed bool unmarshal and the one from a failed JSONSchemaProps unmarshal. Were biasing toward showing the error from JSONSchemaProps unmarshal. Should be returning a custom error that explain a bit more? 

Here, the boolean is trivial, so it's less helpful, but some of the below cases are more interesting.

Reply:
The input types that are acceptable for unmarshaling into bool vs. struct are distinct, and the caller (i.e. the library) guarantees that the input is well-formed, but we can't reliably distinguish up-front as with JSON.

I had thought earlier that it would not be practical to distinguish the error "the input type can't be unmarshaled into a bool or a struct". Thinking about it more, a third unmarshal into `struct{}` would allow you to effectively distinguish whether the input is not a map versus a map with some other problem.

Reply:
The same sort of trick might be possible for "...OrArray" by unmarshaling into a zero-length array: `[0]interface{}`. The CBOR implementation discards excess elements when unmarshaling into arrays like encoding/json does.